File: Medium/2140. Solving Questions With Brainpower/Solution.cpp
class Solution {
public:
    long long dp[100001];

    long long solve(vector<vector<int> > &questions, int i = 0) {
        if(i >= (int) questions.size()) return 0;

        auto & ret = dp[i];
        if(~ret) return ret;

        ret = solve(questions, i + 1);
        ret = max(ret, questions[i][0] + solve(questions, i + questions[i][1] + 1));

        return ret;
    }

    long long mostPoints(vector<vector<int>>& questions) {
        memset(dp, -1, sizeof dp);
        return solve(questions);
    }
};

File: Medium/2140. Solving Questions With Brainpower/Solution.java
class Solution {
    long[] dp = new long[100001];

    long solve(int[][] questions, int i) {
        if(i >= questions.length) return 0;
        if(dp[i] != -1) return dp[i];

        long ret = solve(questions, i + 1);
        ret = Math.max(ret, questions[i][0] + solve(questions, i + questions[i][1] + 1));

        return dp[i] = ret;
    }

    public long mostPoints(int[][] questions) {
        Arrays.fill(dp, -1);
        return solve(questions, 0);
    }
}